esp-hal-common/src/systimer.rs

```diff
@@ -262,3 +263,105 @@ impl<T> Alarm<T, 2> {
         Self { _pd: PhantomData }
     }
 }
+
+#[cfg(feature = "async")]
+mod asynch {
+    use core::{
+        pin::Pin,
+        task::{Context, Poll},
+    };
+
+    use embassy_sync::waitqueue::AtomicWaker;
+    use procmacros::interrupt;
+
+    use super::*;
+
+    const NUM_ALARMS: usize = 3;
+
+    const INIT: AtomicWaker = AtomicWaker::new();
+    static WAKERS: [AtomicWaker; NUM_ALARMS] = [INIT; NUM_ALARMS];
+
+    pub(crate) struct AlarmFuture<'a, const N: u8> {
+        phantom: PhantomData<&'a Alarm<Periodic, N>>,
+    }
+
+    impl<'a, const N: u8> AlarmFuture<'a, N> {
+        pub(crate) fn new(alarm: &'a Alarm<Periodic, N>) -> Self {
+            alarm.interrupt_clear();
+            alarm.enable_interrupt(true);
+
+            Self {
+                phantom: PhantomData,
+            }
+        }
+
+        fn event_bit_is_clear(&self) -> bool {
+            let r = unsafe { &*crate::peripherals::SYSTIMER::PTR }
+                .int_ena
+                .read();
+
+            match N {
+                0 => r.target0_int_ena().bit_is_clear(),
+                1 => r.target1_int_ena().bit_is_clear(),
+                2 => r.target2_int_ena().bit_is_clear(),
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    impl<'a, const N: u8> core::future::Future for AlarmFuture<'a, N> {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
+            WAKERS[N as usize].register(ctx.waker());
+
+            if self.event_bit_is_clear() {
+                Poll::Ready(())
+            } else {
+                Poll::Pending
+            }
+        }
+    }
+
+    impl<const CHANNEL: u8> embedded_hal_async::delay::DelayUs for Alarm<Periodic, CHANNEL> {
+        async fn delay_us(&mut self, us: u32) {
+            let period = MicrosDurationU32::from_ticks(us);
+            self.set_period(period);
+
+            AlarmFuture::new(self).await;
+        }
+
+        async fn delay_ms(&mut self, ms: u32) {
+            for _ in 0..ms {
+                self.delay_us(1000).await;
+            }
+        }
+    }
+
+    #[interrupt]
+    fn SYSTIMER_TARGET0() {
```

Comment:
@jessebraham  This change broke embassy. The interrupt handlers may already be defined by https://github.com/esp-rs/esp-hal/blob/841830dd7fd95d6672efabe185650182b87abb91/esp-hal-common/src/embassy/time_driver_systimer.rs#L68